internal/compiler/passes/remove_return.rs

```diff
@@ -0,0 +1,378 @@
+                            base: load.clone(),
+                            name: FIELD_CONDITION.into(),
+                        }
+                        .into(),
+                        true_expr: if has_value {
+                            Expression::StructFieldAccess {
+                                base: load.clone(),
+                                name: FIELD_ACTUAL.into(),
+                            }
+                        } else {
+                            Expression::default_value_for_type(ty)
+                        }
+                        .into(),
+                        false_expr: if has_return_value {
+                            Expression::StructFieldAccess {
+                                base: load.clone(),
+                                name: FIELD_RETURNED.into(),
+                            }
+                        } else {
+                            Expression::default_value_for_type(ty)
+                        }
+                        .into(),
+                    },
+                ])
+            }
+        }
+    }
+
+    fn into_return_object(self, ty: &Type, ret_ty: &Type) -> Expression {
+        match self {
+            ExpressionResult::Just(e) => make_struct(
+                [
+                    (FIELD_CONDITION, Type::Bool, Expression::BoolLiteral(true)),
+                    (
+                        FIELD_RETURNED.into(),
+                        ret_ty.clone(),
+                        Expression::default_value_for_type(ret_ty),
+                    ),
+                    (FIELD_ACTUAL, e.ty(), e),
+                ]
+                .into_iter(),
+            ),
+            ExpressionResult::MaybeReturn {
+                pre_statements,
+                condition,
+                returned_value,
+                actual_value,
+            } => {
+                let o = Expression::Condition {
+                    condition: condition.into(),
+                    true_expr: ExpressionResult::Just(
+                        actual_value.unwrap_or_else(|| Expression::default_value_for_type(ty)),
+                    )
+                    .into_return_object(ty, ret_ty)
+                    .into(),
+                    false_expr: ExpressionResult::Return(returned_value)
+                        .into_return_object(ty, ret_ty)
+                        .into(),
+                };
+                codeblock_with_expr(pre_statements, o)
+            }
+            ExpressionResult::Return(r) => make_struct(
+                [(FIELD_CONDITION.into(), Type::Bool, Expression::BoolLiteral(false))]
+                    .into_iter()
+                    .chain(r.map(|r| (FIELD_RETURNED, ret_ty.clone(), r)))
+                    .chain((!matches!(ty, Type::Void | Type::Invalid)).then(|| {
+                        (FIELD_ACTUAL.into(), ty.clone(), Expression::default_value_for_type(ty))
+                    })),
```

Comment:
Took me a few rounds to see that the conditions here (r being `Some` and `ty` not being `Void`/`Invalid`) must implicitly match with the `has_value` and `has_return_value` on the caller side. I admit I don't know though how to make that easier to read. Doesn't make sense to pass additional `bool` from the caller to this function - semi-redundant with the `Type` parameters.